packages/auto-tls/src/auto-tls.ts

```diff
@@ -0,0 +1,200 @@
+  private renewTimeout?: ReturnType<typeof setTimeout>
+
+  constructor (components: AutoTLSComponents, init: AutoTLSInit = {}) {
+    this.log = components.logger.forComponent('libp2p:certificate-manager')
+    this.addressManager = components.addressManager
+    this.privateKey = components.privateKey
+    this.peerId = components.peerId
+    this.events = components.events
+    this.forgeEndpoint = init.forgeEndpoint ?? 'registration.libp2p.direct'
+    this.forgeDomain = init.forgeDomain ?? 'libp2p.direct'
+    this.acmeDirectory = init.acmeDirectory ?? 'https://acme-v02.api.letsencrypt.org/directory'
+    this.timeout = init.timeout ?? 10000
+    this.renewThreshold = init.renewThreshold ?? 60000
+    this.clientAuth = new ClientAuth(this.privateKey)
+    this.started = false
+    this.fetching = false
+    this.fetchCertificates = debounce(this._fetchCertificates.bind(this), init.delay ?? 5000)
+  }
+
+  get [serviceDependencies] (): string[] {
+    return [
+      '@libp2p/identify'
+    ]
+  }
+
+  async start (): Promise<void> {
+    if (this.started) {
+      return
+    }
+
+    this.events.addEventListener('self:peer:update', this.fetchCertificates)
+    this.shutdownController = new AbortController()
+    this.started = true
+  }
+
+  async stop (): Promise<void> {
+    this.events.removeEventListener('self:peer:update', this.fetchCertificates)
+    this.shutdownController?.abort()
+    clearTimeout(this.renewTimeout)
+    await stop(this.fetchCertificates)
+    this.started = false
+  }
+
+  private _fetchCertificates (): void {
+    if (this.fetching || this.certificate != null) {
+      this.log('already fetching or already have a certificate')
+      return
+    }
+
+    const addresses = this.addressManager
+      .getAddresses()
+      .filter(ma => !isPrivate(ma) && !isLoopback(ma) && (
+        TCP.exactMatch(ma) ||
+        WebSockets.exactMatch(ma) ||
+        WebSocketsSecure.exactMatch(ma) ||
+        QUICV1.exactMatch(ma) ||
+        WebTransport.exactMatch(ma)
+      ))
+
+    if (addresses.length === 0) {
+      this.log('not fetching certificate as we have no public addresses')
+      return
+    }
+
+    this.fetching = true
+
+    this.fetchCertificate(addresses, {
+      signal: AbortSignal.timeout(this.timeout)
+    })
+      .catch(err => {
+        this.log.error('error fetching certificates %e', err)
+      })
+      .finally(() => {
+        this.fetching = false
+      })
+  }
+
+  private async fetchCertificate (mulitaddrs: Multiaddr[], options?: AbortOptions): Promise<void> {
+    this.log('fetching certificate')
+
+    // TODO: handle rate limit errors like "too many new registrations (10) from this IP address in the last 3h0m0s, retry after 2024-11-01 09:22:38 UTC: see https://letsencrypt.org/docs/rate-limits/#new-registrations-per-ip-address"
+
+    const base36EncodedPeer = base36.encode(this.peerId.toCID().bytes)
+    const domain = `${base36EncodedPeer}.${this.forgeDomain}`
```

Comment:
Need to be a wildcard:

```suggestion
    const domain = `*.${base36EncodedPeer}.${this.forgeDomain}`
```